File: dashboard/lib/formatters.ts
// lib/formatters.ts

/**
 * Formatea un número como moneda ARS
 * Ejemplo: 903849.50 → $ 903.849,50
 */
export function formatCurrency(value: number): string {
  if (value === undefined || value === null || isNaN(value)) {
    return '$ 0,00'
  }
  
  return new Intl.NumberFormat('es-AR', {
    style: 'currency',
    currency: 'ARS',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(value)
}

/**
 * Formatea un número como porcentaje
 * Ejemplo: 34.78 → 34,78%
 */
export function formatPercentage(value: number): string {
  if (value === undefined || value === null || isNaN(value)) {
    return '0,00%'
  }
  
  return new Intl.NumberFormat('es-AR', {
    style: 'percent',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(value / 100)
}

/**
 * Formatea un número con separador de miles
 * Ejemplo: 903849 → 903.849
 */
export function formatNumber(value: number): string {
  if (value === undefined || value === null || isNaN(value)) {
    return '0'
  }
  
  return new Intl.NumberFormat('es-AR').format(value)
}

/**
 * Formatea una fecha
 * Ejemplo: 2026-07-26 → 26/07/2026
 */
export function formatDate(dateStr: string): string {
  if (!dateStr) return 'N/A'
  try {
    const date = new Date(dateStr)
    return date.toLocaleDateString('es-AR')
  } catch {
    return dateStr
  }
}

File: backend/app/utils/formatters.py
# app/utils/formatters.py
"""
Funciones de formateo para números y fechas
"""

def format_number(value: float, decimals: int = 2) -> str:
    """
    Formatea un número con separador de miles (.) y decimales (,)
    Ejemplo: 1234567.89 → "1.234.567,89"
    """
    if value is None:
        return "0,00"
    
    # Formatear con separador de miles y decimales
    if decimals == 0:
        return f"{int(value):,}".replace(",", ".")
    
    parte_entera = int(abs(value))
    parte_decimal = round(abs(value) - parte_entera, decimals)
    
    # Formatear parte entera con separador de miles
    entera_formateada = f"{parte_entera:,}".replace(",", ".")
    
    # Formatear parte decimal con 2 dígitos
    decimal_str = f"{parte_decimal:.{decimals}f}"[2:]
    
    signo = "-" if value < 0 else ""
    return f"{signo}{entera_formateada},{decimal_str}"


def format_currency(value: float) -> str:
    """
    Formatea un número como moneda (ARS)
    Ejemplo: 1234567.89 → "$ 1.234.567,89"
    """
    if value is None:
        return "$ 0,00"
    return f"$ {format_number(value)}"


def format_percentage(value: float, decimals: int = 2) -> str:
    """
    Formatea un número como porcentaje
    Ejemplo: 37.36 → "37,36%"
    """
    if value is None:
        return "0,00%"
    return f"{format_number(value, decimals)}%"